pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,337 @@
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.V(2).Infof("Couldn't get object from tombstone %#v", obj)
+			return
+		}
+		shard, ok = tombstone.Obj.(*tenancyv1alpha1.WorkspaceShard)
+		if !ok {
+			klog.V(2).Infof("Tombstone contained object that is not a WorkspaceShard: %#v", obj)
+			return
+		}
+	}
+	klog.Infof("handling removed shard %q", shard.Name)
+	workspaces, err := c.workspaceIndexer.ByIndex(currentShardIndex, shard.Name)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	for _, workspace := range workspaces {
+		key, err := cache.MetaNamespaceKeyFunc(workspace)
+		if err != nil {
+			runtime.HandleError(err)
+			return
+		}
+		klog.Infof("queuing orphaned workspace %q", key)
+		c.queue.Add(key)
+	}
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting Workspace controller")
+	defer klog.Info("Shutting down Workspace controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	klog.Infof("processing key %q", key)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	namespace, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("invalid key: %q: %v", key, err)
+		return nil
+	}
+	if namespace != "" {
+		klog.Errorf("namespace %q found in key for cluster-wide Workspace object", namespace)
+		return nil
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.workspaceLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	previous := obj
+	obj = obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Status, obj.Status) {
+		oldData, err := json.Marshal(tenancyv1alpha1.Workspace{
+			Status: previous.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal old data for workspace %q|%q/%q: %v", clusterName, namespace, name, err)
```

Comment:
Are we wrapping errors (%w) or does it not matter?

Reply:
I will add a commit to enforce this.

Reply:
https://github.com/kcp-dev/kcp/pull/255

Reply:
Hilariously, the `golangci-lint` `errorlint` check did not fail on this :facepalm: